### PR TITLE
[#8431] InstructorFeedbackEditPage: include question type help in lightbox

### LINKED ISSUE #8431: InstructorFeedbackEditPage: include question type help in lightbox
**Description of feature/enhancement**
image
At the moment, clicking the info button (shown above) when trying to add a new question in the Edit Feedback Session screen opens `instructorHelp.jsp#fbQuestionTypes` in a new tab. While this is workable, it might be more beneficial to instructors' workflow to have information on question types available on the same page, for example in a lightbox as shown below:
peek 2018-02-10 15-45

**Justification**
Instructors will benefit from not having to navigate away from the page to learn more about question types. The current info button opens a new tab in the background. This impedes workflow as the user must navigate away from the page, read the information, then navigate back to the page to continue. Users may also get distracted by the (vey comprehensive) information available on `instructorHelp.jsp`.

Moreover, we are already using lightboxes in other parts of the site when displaying additional information; we should standardize what behaviour is expected when the user clicks on the information icon (i.e. lightbox vs. new tab).

To be fair, the question type help is quite long and elaborate, and may not fit well in a lightbox -- we may want to look into including a condensed version instead.

### PR DESCRIPTION
**Pull request by thikhinab**
Fixes #8431

**Outline of Solution**
Updated the front-end with buttons corresponding to each question type to open a new tab with helpful information (specific to the question type) as per discussion in the issue thread.

**UI**
Screencast from 2022-08-25 04 18 49

### CODE REVIEW

**Comment by damithc** [comment]
I'm OK with the proposed UX. But other views welcome.

**Comment by thikhinab** [comment]
@damithc 

**Alternative UX**
1. The dropdown list of questions would have a `tooltip` with a brief description of the type of question.
2. Once a question is selected and the `question-edit-from` is opened,  we can have ` information button` which would open a modal with all the specific details on setting up the particular type of question.

**Comment by damithc** [comment]
> **Alternative UX**
> 
> 1. The dropdown list of questions would have a `tooltip` with a brief description of the type of question.
> 2. Once a question is selected and the `question-edit-from` is opened,  we can have ` information button` which would open a modal with all the specific details on setting up the particular type of question.

That's a good alternative @thikhinab I suspect instructors would want to see more than a brief description before making the decision on which question type to use though.

**Comment by NicolasCwy** [comment]
@thikhinab Has this PR been completed? You can add labels on the side so we know if someone should be reviewing this or if this PR is still ongoing

**Comment by thikhinab** [comment]
@NicolasCwy PR is completed according to the current proposal. If there are any other suggestions I can work on them. 
I don't think I have the rights to add labels.
image

**Comment by ziqing26** [comment]
@NicolasCwy @wkurniawan07 For your kind attention, thank you!

File: src/web/app/components/adding-question-panel/adding-question-panel.component.html
Comment (NicolasCwy) [inline src/web/app/components/adding-question-panel/adding-question-panel.component.html:15]: Please look into the indentation for the changes in this file and align them with their opening tag. Any nesting of tags should also be one level deeper/ on a new line (i tags)

File: src/web/app/components/adding-question-panel/adding-question-panel.component.html
Comment (NicolasCwy) [inline src/web/app/components/adding-question-panel/adding-question-panel.component.html:10]: I believe the `*ngIf="questionType !== FeedbackQuestionType.CONSTSUM"` is not needed for the button element since it is already wrapped in a div that ensures that the question type is not constant sum

File: src/web/app/components/adding-question-panel/adding-question-panel.component.spec.ts
Comment (NicolasCwy) [inline src/web/app/components/adding-question-panel/adding-question-panel.component.spec.ts:43]: Can we add an assertion statement to check that Object.keys(component.FeedbackQuestionType)[0] is not constant sum question type, in case any changes are made in the future to this file, to ensure that there is an actual corresponding search query

File: src/web/app/components/teammates-common/question-type-id.pipe.ts
Comment (NicolasCwy) [inline src/web/app/components/teammates-common/question-type-id.pipe.ts:8]: Perhaps we can rename this file to question-type-help-path to be more descriptive?

File: src/web/app/components/teammates-common/question-type-id.pipe.ts
Comment (NicolasCwy) [inline src/web/app/components/teammates-common/question-type-id.pipe.ts:37]: Should the default case ever execute? We could throw an error here instead of returning a dummy value
Have not tried this but found `invalidPipeArgumentError` to be a possible one

**Review by NicolasCwy — CHANGES_REQUESTED** [review]
Sorry for the late review. Added some of my comments on changes that could be made

**Comment by ziqing26** [comment]
@thikhinab For your kind attention on this PR, feel free to let us know if you want to put this on hold first.

**Comment by thikhinab** [comment]
@ziqing26 Sorry for the delay, I will try to make changes this week

File: src/web/app/components/adding-question-panel/adding-question-panel.component.html
Comment (thikhinab) [reply src/web/app/components/adding-question-panel/adding-question-panel.component.html:15]: Will do!

File: src/web/app/components/adding-question-panel/adding-question-panel.component.html
Comment (thikhinab) [reply src/web/app/components/adding-question-panel/adding-question-panel.component.html:10]: You are right, will make the change.

File: src/web/app/components/adding-question-panel/adding-question-panel.component.spec.ts
Comment (thikhinab) [reply src/web/app/components/adding-question-panel/adding-question-panel.component.spec.ts:43]: That is good idea. Will add!

File: src/web/app/components/teammates-common/question-type-id.pipe.ts
Comment (thikhinab) [reply src/web/app/components/teammates-common/question-type-id.pipe.ts:8]: Thats true, will update

File: src/web/app/components/teammates-common/question-type-id.pipe.ts
Comment (thikhinab) [reply src/web/app/components/teammates-common/question-type-id.pipe.ts:37]: @NicolasCwy Can you advise me on how to add assertions? Should I use an external library?

**Comment by thikhinab** [comment]
@ziqing26 I accidentally removed @wkurniawan07 review request, can you re-request it?

**Comment by damithc** [comment]
@TEAMMATES/active-mentors reviews needed for this PR.

File: src/web/app/components/adding-question-panel/adding-question-panel.component.html
Comment (zhaojj2209) [reply src/web/app/components/adding-question-panel/adding-question-panel.component.html:15]: Reminder to fix the indentation here! Specifically, line 12 seems to have 6 extra spaces. Personally, I think lines 10 and 11 can also be combined into 1 line.

**Review by zhaojj2209 — APPROVED** [review]
LGTM

File: src/web/app/components/adding-question-panel/adding-question-panel.component.html
Comment (thikhinab) [reply src/web/app/components/adding-question-panel/adding-question-panel.component.html:14]: @zhaojj2209 just updated the changes!

**Comment by thikhinab** [comment]
Please let me know if any updates are needed, thank you!

**Comment by thikhinab** [comment]
@NicolasCwy Please let me know if there any updates

**Review by hhdqirui — COMMENTED** [review]
Has this (https://github.com/TEAMMATES/teammates/pull/11949#discussion_r987150466) really been addressed yet?

**Comment by thikhinab** [comment]
@hhdqirui Awaiting advice, would you be able to direct me?

**Comment by thikhinab** [comment]
@zhaojj2209 I added an assertion using `console.assert`. I believe this is not the standard way to do it and does not conform with the `es-lint`. Can you please advise on how I should proceed?

**Comment by zhaojj2209** [comment]
> @zhaojj2209 I added an assertion using `console.assert`. I believe this is not the standard way to do it and does not conform with the `es-lint`. Can you please advise on how I should proceed?

Sorry for the late response. TBH I have no idea why the suggestion was to add an assert statement when the proper way to handle it would be to throw an exception. Throwing an error with a message like `Unknown question type` would be better.

**Comment by damithc** [comment]
> Sorry for the late response. TBH I have no idea why the suggestion was to add an assert statement when the proper way to handle it would be to throw an exception. Throwing an error with a message like `Unknown question type` would be better.

The standard way to handle the assumption "this can never happen" is to use an assert, so that if someone broke that assumption, the code will crash. However, we have to be careful when doing that in production systems. 
If the assumption is broken, will the assertion failure be discovered during testing, before merging? In that case, an assertion is fine.
But if there is a chance that the assumption breakage can go undiscovered into production, an exception might be more appropriate, to handle the situation more gracefully. Even then, the problem should be reported in a unmistakable way so that it gets noticed immediately (after all, it indicates there is a potential bug in the code).

**Comment by zhaojj2209** [comment]
> The standard way to handle the assumption "this can never happen" is to use an assert, so that if someone broke that assumption, the code will crash. However, we have to be careful when doing that in production systems. If the assumption is broken, will the assertion failure be discovered during testing, before merging? In that case, an assertion is fine. But if there is a chance that the assumption breakage can go undiscovered into production, an exception might be more appropriate, to handle the situation more gracefully. Even then, the problem should be reported in a unmistakable way so that it gets noticed immediately (after all, it indicates there is a potential bug in the code).

Thanks for the clarification Prof!
Unlike Java, there is no standard `assert` in JavaScript; `console.assert` will merely write a message to the console, so it will be easy for the assertion to go undetected. Therefore I think it would be better to throw an error instead so that any errors can be easily caught.

**Comment by thikhinab** [comment]
@damithc @zhaojj2209 Thank you for the clarification.
Will make the necessary changes.

**Comment by hhdqirui** [comment]
Hi @thikhinab kind reminder on this PR.

**Comment by thikhinab** [comment]
@hhdqirui  Sorry for the delay.
@zhaojj2209  I need some advice. Throwing an error does not give the desired effect of warning the developer as discussed earlier. Similar to `console.assert`, the error message simply gets logged in the console.

Console:
image

Snippet of code at `QuestionTypeHelpPathPipe` I used for testing:
```typescript
  transform(type: FeedbackQuestionType): string | Error {
    switch (type) {
      case FeedbackQuestionType.MCQ:
        return QuestionsSectionQuestions.SINGLE_ANSWER_MCQ;
      case FeedbackQuestionType.CONTRIB:
        return QuestionsSectionQuestions.CONTRIBUTION;
      
     ....

      default:
        throw new Error('Invalid Question Type');
    }
```

**Comment by zhaojj2209** [comment]
@thikhinab Though they seem similar, there are still advantages to throwing errors instead of using `console.assert`:
- Code continues executing after `console.assert`, but stops executing when an error is thrown
- `console.assert` cannot be caught by testcases, whereas thrown errors will cause relevant tests to fail
- Thrown errors can be caught and handled such that the error shows on the UI (e.g. in error toast notifications)

The error is currently showing only on the console because there is no code to handle the thrown error; for this use case, there is no need to have the error show on the frontend. Nonetheless, any tests that make use of this pipe will fail if the default case is ever reached.

Hope this clears up some doubts, though I may not be 100% accurate with the facts here so do read up more on error handling in JS if you're interested to learn more. In the meantime, do proceed with the change.

**Comment by thikhinab** [comment]
@zhaojj2209 Understood, I will proceed as advised. Thank you!

File: src/web/app/components/teammates-common/question-type-help-path.pipe.ts
Comment (hhdqirui) [inline src/web/app/components/teammates-common/question-type-help-path.pipe.ts:36]: This line is incorrect. It should be `QuestionsSectionQuestions.POINTS_RECIPIENTS`. Please make the changes.

**Review by hhdqirui — CHANGES_REQUESTED** [review]
Apart from the one comment. Everything else looks good to me.

File: src/web/app/components/teammates-common/question-type-help-path.pipe.ts
Comment (thikhinab) [reply src/web/app/components/teammates-common/question-type-help-path.pipe.ts:36]: Thanks for pointing it out!

**Comment by thikhinab** [comment]
@hhdqirui I requested review from you. Other reviewers got automatically removed, you can add them under your discretion.

**Comment by hhdqirui** [comment]
@thikhinab Hi please also update the snapshot test. Thanks!

**Review by hhdqirui — APPROVED** [review]
LGTM!

**Pull request merged**
